### PR TITLE
feat: add support for custom labels

### PR DESCRIPTION
**Pull request by manishdangi98**
Add support for adding additional OTel attributes to the context. 
These attributes are added to metrics and spans generated for the query.

### CODE REVIEW

File: query.go [hunk 1]
```diff
@@ -24,11 +26,41 @@ func skippedQueryContext(_ context.Context, _ string, _ []driver.NamedValue) (dr
...
+func AddTracerLabels(ctx context.Context, labels ...attribute.KeyValue) context.Context {
+	return addLabels(ctx, tracerCtxKey, labels...)
+}
+
 // queryStats records metrics for query.
 func queryStats(r methodRecorder, method string) queryContextFuncMiddleware {
 	return func(next queryContextFunc) queryContextFunc {
 		return func(ctx context.Context, query string, args []driver.NamedValue) (result driver.Rows, err error) {
-			end := r.Record(ctx, method)
+			labels := getLabels(ctx, meterCtxKey)
```
Comment (mabdh) [inline query.go]: can we get the information/table name from `query` instead?

File: query.go [hunk 2]
```diff
@@ -24,11 +26,41 @@ func skippedQueryContext(_ context.Context, _ string, _ []driver.NamedValue) (dr
...
+}
+
+func AddTracerLabels(ctx context.Context, labels ...attribute.KeyValue) context.Context {
+	return addLabels(ctx, tracerCtxKey, labels...)
+}
+
 // queryStats records metrics for query.
 func queryStats(r methodRecorder, method string) queryContextFuncMiddleware {
 	return func(next queryContextFunc) queryContextFunc {
 		return func(ctx context.Context, query string, args []driver.NamedValue) (result driver.Rows, err error) {
```
Comment (mabdh) [inline query.go:50]: we could pass the operation as an arg too

File: query.go [hunk 3]
```diff
@@ -24,11 +26,41 @@ func skippedQueryContext(_ context.Context, _ string, _ []driver.NamedValue) (dr
 	return nil, driver.ErrSkip
 }
 
+// add Custom Lables to metrics and traces
+type ctxKey string
+
+const (
+	meterCtxKey  ctxKey = "meterCtxKey"
```
Comment (mabdh) [inline query.go]: Possible to use `struct{}` here? 
ref: https://forum.golangbridge.org/t/the-way-to-set-a-context-key-and-value/16311

Comment (manishdangi98) [reply query.go]: We can use struct here but we will have to define 2 different types for the context keys like this
```
type tracerCtxKey struct{}
type meterCtxKey struct{}

context.WithValue(ctx, tracerCtxKey{}, attrs)
```

I decided to use a single type just to have some association between the keys, and also the keys and the type is not exported directly so we can be sure that it won't be misused.

But looking at it now I think it is better to stick to `struct{}` as it's not a built-in type. I'll update this, thanks!

Comment (mabdh) [reply query.go]: Please use struct{} for the key


https://pkg.go.dev/context#WithValue

File: query.go [hunk 4]
```diff
@@ -24,11 +26,41 @@ func skippedQueryContext(_ context.Context, _ string, _ []driver.NamedValue) (dr
...
+	tracerCtxKey ctxKey = "traceCtxKey"
```
Comment (mabdh) [inline query.go]: I dont think we need to separate tracer and meter attribute here. Why it is different between tracer and meter?

File: query.go [hunk 5]
```diff
@@ -24,11 +26,41 @@ func skippedQueryContext(_ context.Context, _ string, _ []driver.NamedValue) (dr
...
+	}
+	return labels
+}
+
+func addLabels(ctx context.Context, key ctxKey, labels ...attribute.KeyValue) context.Context {
+	new := append(getLabels(ctx, key), labels...)
+	return context.WithValue(ctx, key, new)
+}
+
+func AddMeterLabels(ctx context.Context, labels ...attribute.KeyValue) context.Context {
```
Comment (mabdh) [inline query.go]: We should not call it labels but we are passing attribute.
We have to change the method name to something like WithCustomAtrributes or WithAdditionalAttributes

File: query.go (hunk 4 above)
Comment (manishdangi98) [reply query.go]: Do you mean we should have 1 function to add attributes to both metrics and traces? We will loose the ability to set different custom attributes for meter and tracer, but that sounds okay for most cases.

Or should I remove it completely as it's not being used by us?

I added `AddTracerLabels` function just for the sake of completeness mostly.

File: query.go (hunk 3 above)
Comment (manishdangi98) [reply query.go]: Pushed changes for this

File: query.go (hunk 5 above)
Comment (manishdangi98) [reply query.go]: Done

File: query.go (hunk 4 above)
Comment (mabdh) [reply query.go]: Lets not distinguish them. The attributes should be reusable for both.

File: query.go [hunk 6]
```diff
@@ -24,11 +26,30 @@ func skippedQueryContext(_ context.Context, _ string, _ []driver.NamedValue) (dr
...
+	attrs, ok := ctx.Value(attributeCtxKey{}).([]attribute.KeyValue)
+	if !ok || attrs == nil {
+		attrs = make([]attribute.KeyValue, 0)
+	}
+	return attrs
+}
+
+// AddCutomAttributes adds additional OTel attributes to the context.
+// These attributes are added to metrics and spans generated for the query.
+func AddCustomAttributes(ctx context.Context, labels ...attribute.KeyValue) context.Context {
```
Comment (mabdh) [inline query.go]:
```suggestion
func WithCustomAttributes(ctx context.Context, labels ...attribute.KeyValue) context.Context {
```

The naming could follow how https://pkg.go.dev/context#WithValue package does

**Pull request merged**